components/encryption/src/io.rs

```diff
@@ -203,23 +219,19 @@ impl<W: Write> Write for EncrypterWriter<W> {
                 ),
             ));
         }
-        self.writer.write_all(&encrypt_buffer[0..bytes])?;
+        let writer = self.writer.as_mut().unwrap();
+        writer.write_all(&encrypt_buffer[0..bytes])?;
         Ok(bytes)
     }
 
     fn flush(&mut self) -> IoResult<()> {
-        self.writer.flush()
+        let writer = self.writer.as_mut().unwrap();
+        writer.flush()
     }
 }
 
 impl<W: Write> Drop for EncrypterWriter<W> {
     fn drop(&mut self) {
-        let mut encrypt_buffer = vec![0; self.block_size];
-        let bytes = self.crypter.finalize(&mut encrypt_buffer).unwrap();
-        if bytes != 0 {
-            // The EncrypterWriter current only support crypters that always return the same amount
-            // of data. This is true for CTR mode.
-            panic!("unsupported encryption");
-        }
+        self.do_finalize();
```

Comment:
it looks like if `finalize()` is called, `do_finalize()` (as well as `crypter.finalize()`) will be called twice (the second time in `drop()`). can we avoid calling `crypter.finalize()` a second time?

Reply:
God catch.